### PR TITLE
[ShapeOpt] Geometric constraint: first version of packaging

### PR DESCRIPTION
**Pull request by armingeiser**
This PR introduces a simple version of packaging response functions:
* plane
* bounding mesh

Both work by accumulating the nodal violations into a single response function, using a square sum of all violations. This means the constraint will only be satisfied in a weak sense. See e.g. the small violations (blue areas) in the animation below.

kratos_packaging

### CODE REVIEW

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function.py:1]: Header is missing. Also in the other files.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+import time as timer
+
+import KratosMultiphysics
+from KratosMultiphysics import Logger
+
+from .packaging_response_base import PackagingResponseBase
+from .. import custom_math as cm
+
+import numpy as np
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py]: Not used here.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        self.identifier = identifier
+
+        self.response_settings = response_settings
+        self.model = model
+        self.model_part_needs_to_be_imported = False
+
+        model_part_name = response_settings["model_part_name"].GetString()
+        input_type = response_settings["model_import_settings"]["input_type"].GetString()
+        if input_type == "mdpa":
+            self.model_part = self.model.CreateModelPart(model_part_name, 2)
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: Why domain size 2?

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            domain_size = response_settings["domain_size"].GetInt()
+            if domain_size not in [2, 3]:
+                raise Exception("MassResponseFunction: Invalid 'domain_size': {}".format(domain_size))
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: Is this somehow linked to the Mass Response function? In general, I don't quite get the logic of this if section.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            raise Exception("Other model part input options are not yet implemented.")
+
+        self.signed_distances = None
+        self.directions = None
+
+        self.value = None
+
+        self.gradient = {}
+
+        method_class = globals()[self.response_settings["method"].GetString()]
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: What does this "globals()" do here?

File: applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py [hunk 6]
```diff
@@ -1,93 +1,99 @@
-# ==============================================================================
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py]: Why does it indicate a change of the whole file?

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+        if not self.directions or not self.signed_distances:
+            self._CalculateProjectedDistances()
+
+        value = 0.0
+        for i in range(len(self.signed_distances)):
+
+            _direction = np.array(self.directions[i*3:i*3+3])
+
+            value += self.method.GetValue(self.signed_distances[i], _direction)
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: You need to explain to me, how this thing with the method works.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
+from __future__ import print_function, absolute_import, division
+
+import time as timer
+
+import KratosMultiphysics
+from KratosMultiphysics import Logger
+
+from .response_function import ResponseFunctionBase
+
+import numpy as np
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: We need to get rid of this dependency, which is not difficult, I guess.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/mesh_packaging_response.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    def __init__(self, identifier, response_settings, model):
+
+        super().__init__(identifier, response_settings, model)
+
+        self.packaging_model_part = None
+        self.packaging_model_part_needs_to_be_imported = False
+
+        packaging_model_part_name = response_settings["packaging_model_part_name"].GetString()
+        input_type = response_settings["packaging_model_import_settings"]["input_type"].GetString()
+        if input_type == "mdpa":
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/mesh_packaging_response.py]: Here, also a vrml input should be possible.

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h [hunk 10]
```diff
@@ -97,16 +101,22 @@ class GeometryUtilities
 
     // --------------------------------------------------------------------------
     void ComputeUnitSurfaceNormals()
+    {
+        _ComputeUnitSurfaceNormals(mrModelPart);
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]: Why this change?

File: applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp [hunk 11]
```diff
@@ -168,6 +168,7 @@ void  AddCustomUtilitiesToPython(pybind11::module& m)
         .def("ComputeUnitSurfaceNormals", &GeometryUtilities::ComputeUnitSurfaceNormals)
         .def("ProjectNodalVariableOnUnitSurfaceNormals", &GeometryUtilities::ProjectNodalVariableOnUnitSurfaceNormals)
         .def("ExtractBoundaryNodes", &GeometryUtilities::ExtractBoundaryNodes)
+        .def("ComputeProjectedDistancesToBoundingModelPart", &GeometryUtilities::ComputeProjectedDistancesToBoundingModelPart)
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp]: What about shortening it to "ComputeDistancesToBoundingModelPart"? Note that the distances itself are not projected, so it might be even misleading.

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h [hunk 12]
```diff
@@ -196,6 +206,68 @@ class GeometryUtilities
...
+        // Type definitions for tree-search
+        typedef Node < 3 > NodeType;
+        typedef NodeType::Pointer NodeTypePointer;
+        typedef std::vector<NodeType::Pointer> NodeVector;
+        typedef std::vector<NodeType::Pointer>::iterator NodeIterator;
+        typedef std::vector<double>::iterator DoubleVectorIterator;
+        typedef Bucket< 3, NodeType, NodeVector, NodeTypePointer, NodeIterator, DoubleVectorIterator > BucketType;
+        typedef Tree< KDTreePartition<BucketType> > KDTree;
+
+        size_t bucket_size = 1000;
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]: Why bucket size 1000?

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h [hunk 13]
```diff
@@ -196,6 +206,68 @@ class GeometryUtilities
     	KRATOS_CATCH("");
     }
 
+    // --------------------------------------------------------------------------
+    void ComputeProjectedDistancesToBoundingModelPart(
+        ModelPart& rBoundingModelPart,
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h:205]: The function assumes that `rBoundingModelPart` is a surface part. So maybe it makes sense to include a quick check in the beginning, to make sure it does not contain volume elements.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    def _GetGradient(self, distance, normal):
+        factor = np.exp(distance)
+        gradient = [
+            normal[0] * factor,
+            normal[1] * factor,
+            normal[2] * factor
+        ]
+        return gradient
+
+class PackagingResponseBase(ResponseFunctionBase):
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:16]: From a general perspective: Doing it through the definition of a single response function already implies, that all distances are summed and combined to a single compromise function. This excluded an explicit treatment of all constraints separately. But anyways Kratos cannot handle such a situation right now. But what about "constraint aggregation", like a KS-aggregation. This should be doable in theory, right?

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+        self.plane_origin = []
+        for i in range(0,3):
+            self.plane_origin.append(self.response_settings["plane_origin"][i].GetDouble())
+
+        self.plane_normal = []
+        for i in range(0,3):
+            self.plane_normal.append(self.response_settings["plane_normal"][i].GetDouble())
+
+        self.plane_normal = cm.ScalarVectorProduct(1.0/cm.Norm2(self.plane_normal), self.plane_normal)
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py]: Hmm, I don't quite like the heavy use of the custom_math capabilities here. Simply because we should get rid of of the whole script in favor of standard capabilities. But I understand its necessary here.

**Review by dbaumgaertner — CHANGES_REQUESTED** [review]
Good job @armingeiser! I think its already in a pretty good shape from the functionality point of view. There still some open points that I would like to discuss:

- application of "ValueAndGradient" functionality
- application of custom_match.py
- application of numpy
- possible blockers / obvious efficiency killers / obvious drawbacks (are there even any?)
- vrml input

But this we should do in person. I already left some comments as part of the review to have a basis for discussion. Also, I have some general comments in the following:

- In this PR I like the separation of the response function files into a separate folder and making a module out of it. I makes the development of new response functions more comprehensible.  

- We should make an extra round to clean the code. Right now, a lot of leftovers from debugging are included and also the code style can be improved. It makes sense in this WIP stage. Just to not forget it.

- Can you add a test for each the plane and the mesh packaging? It would be good to quickly test enhancements or changes.

**Comment by dbaumgaertner** [comment]
Again I merged the current master to have the new compile mechanism included.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py (hunk 3 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: This is buffer size. Anyway it is not needed and i removed it.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py (hunk 14 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:16]: Yes this can be done at this stage. In the end, the current implementation is nothing else than a squared-sum aggregation of all violations.

Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:16]: But as discussed i simplified this a bit and removed the unused parts.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py (hunk 15 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py]: I agree, but it is better to use it as long as it is there, instead of implementing dot product and similar again in this file.

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h (hunk 12 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]: fixed - set to 100 as in our other files.

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h (hunk 13 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h:205]: done

File: applications/ShapeOptimizationApplication/python_scripts/custom_ios/wrl_reader.py [hunk 16]
```diff
@@ -15,6 +17,20 @@ def __init__(self, name, nodes, triangles):
...
+    vrml_name = file_name + ".vrml"
+
+    if os.path.isfile(wrl_name) and os.path.isfile(vrml_name):
+        raise Exception("Wrl reader: file {} found with '.wrl' and '.vrml' ending! Please remove one of them!".format(file_name))
+    elif os.path.isfile(wrl_name):
+        return wrl_name
+    elif os.path.isfile(vrml_name):
+        return vrml_name
+    else:
+        raise Exception("Wrl reader: file {} not found with '.wrl' and '.vrml' ending!".format(file_name))
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/custom_ios/wrl_reader.py]:
```suggestion
        raise Exception("Wrl reader: file {} not found with '.wrl' or '.vrml' ending!".format(file_name))
```

File: applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp (hunk 11 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_python/add_custom_utilities_to_python.cpp]: The distance is computed by searching the nearest node and then projecting the distance vector onto the normals of the bounding part. Thats why "projected" is in the name. If you think it is cleaner without, i can remove it.

**Comment by armingeiser** [comment]
@dbaumgaertner this is ready for final review now.

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h [hunk 17]
```diff
@@ -196,6 +200,65 @@ class GeometryUtilities
...
+        typedef NodeType::Pointer NodeTypePointer;
+        typedef std::vector<NodeType::Pointer> NodeVector;
+        typedef std::vector<NodeType::Pointer>::iterator NodeIterator;
+        typedef std::vector<double>::iterator DoubleVectorIterator;
+        typedef Bucket< 3, NodeType, NodeVector, NodeTypePointer, NodeIterator, DoubleVectorIterator > BucketType;
+        typedef Tree< KDTreePartition<BucketType> > KDTree;
+
+        // check if there are elements
+        KRATOS_ERROR_IF(rBoundingModelPart.NumberOfElements() != 0) <<
+            "ComputeProjectedDistancesToBoundingModelPart: Model part should only contain conditions!" << std::endl;
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]:
```suggestion
            "ComputeProjectedDistancesToBoundingModelPart: Model part must only contain conditions!" << std::endl;
```

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h [hunk 18]
```diff
@@ -196,6 +200,65 @@ class GeometryUtilities
     	KRATOS_CATCH("");
     }
 
+    // --------------------------------------------------------------------------
+    void ComputeProjectedDistancesToBoundingModelPart(
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]:
```suggestion
    void ComputeDistancesToBoundingModelPart(
```

What about reducing the function name? This requires also a few adjustments further down the code. I see the meaning of "projected" but actually without further insight, the addition does not provide more information compared to the name without this addition. What do you think?

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h [hunk 19]
```diff
@@ -196,6 +200,65 @@ class GeometryUtilities
...
+        ModelPart& rBoundingModelPart,
+        pybind11::list& rSignedDistances,
+        pybind11::list& rDirections )
+    {
+        KRATOS_TRY;
+
+        // Type definitions for tree-search
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]: Imo the comments are not necessary in this function. They just distract and are prone to be obsolete. What about shortening it as follows:
```
        KRATOS_TRY;

        typedef Node < 3 > NodeType;
        typedef NodeType::Pointer NodeTypePointer;
        typedef std::vector<NodeType::Pointer> NodeVector;
        typedef std::vector<NodeType::Pointer>::iterator NodeIterator;
        typedef std::vector<double>::iterator DoubleVectorIterator;
        typedef Bucket< 3, NodeType, NodeVector, NodeTypePointer, NodeIterator, DoubleVectorIterator > BucketType;
        typedef Tree< KDTreePartition<BucketType> > KDTree;

        KRATOS_ERROR_IF(rBoundingModelPart.NumberOfElements() != 0) <<
            "ComputeProjectedDistancesToBoundingModelPart: Model part should only contain conditions!" << std::endl;

        NodeVector all_bounding_nodes;
        all_bounding_nodes.reserve(rBoundingModelPart.Nodes().size());
        for (ModelPart::NodesContainerType::iterator node_it = rBoundingModelPart.NodesBegin(); node_it != rBoundingModelPart.NodesEnd(); ++node_it)
        {
            all_bounding_nodes.push_back(*(node_it.base()));
        }
        const size_t bucket_size = 100;
        KDTree search_tree(all_bounding_nodes.begin(), all_bounding_nodes.end(), bucket_size);

        GeometryUtilities(rBoundingModelPart).ComputeUnitSurfaceNormals();

        for (auto& r_node : mrModelPart.Nodes()){

            double distance;
            NodeTypePointer p_neighbor = search_tree.SearchNearestPoint(r_node, distance);

            const array_3d delta = r_node.Coordinates() - p_neighbor->Coordinates();
            const array_3d& bounding_normal = p_neighbor->FastGetSolutionStepValue(NORMALIZED_SURFACE_NORMAL);
            const double projected_length = inner_prod(delta, bounding_normal);

            rSignedDistances.append(projected_length);

            rDirections.append(bounding_normal[0]);
            rDirections.append(bounding_normal[1]);
            rDirections.append(bounding_normal[2]);
        }

        KRATOS_CATCH("");
```

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/mesh_packaging_response.py [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+#  KratosShapeOptimizationApplication
+#
+#  License:         BSD License
+#                   license: ShapeOptimizationApplication/license.txt
+#
+#  Main authors:    Geiser Armin, https://github.com/armingeiser
+#
+# ==============================================================================
+import KratosMultiphysics as KM
+
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/mesh_packaging_response.py]: If you don't have comments between the different imports, I would also remove the empty lines in between. So the imports appear as one block.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/mesh_packaging_response.py [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+import KratosMultiphysics.ShapeOptimizationApplication as KSO
+
+from .packaging_response_base import PackagingResponseBase
+from ..custom_ios.wrl_io import WrlIO
+
+class MeshPackagingResponse(PackagingResponseBase):
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/mesh_packaging_response.py]: What about naming the two response functions: `MeshBasedPackaging` and `PlaneBasedPackaging`? I find the current names not very descriptive.

The postfix "Response" is not necessary. Obviously this requires also an adjustment of the settings and the other variables as well as the filename: `plane_based_packaging`.  In the latter is not necessary because it anyways lies in the "response_function"-folder.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+#
+#  License:         BSD License
+#                   license: ShapeOptimizationApplication/license.txt
+#
+#  Main authors:    Geiser Armin, https://github.com/armingeiser
+#
+# ==============================================================================
+
+import time as timer
+
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: Again remove the empty lines.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 23]
```diff
@@ -0,0 +1,157 @@
...
+        self.gradient = {}
+
+        self.infeasible_side = self.response_settings["infeasible_side"].GetBool()
+        self.exponent = 2
+
+    @classmethod
+    def GetDefaultSettings(cls):
+        this_defaults = KM.Parameters("""{
+            "response_type"         : "plane_packaging",
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: I would not set "plane_packaging" here as a default as this is the base class. Instead, leave it completely generic like "UNKNOWN_TYPE" and force the derived classes to specify this setting.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 24]
```diff
@@ -0,0 +1,157 @@
...
+
+        self.value = None
+
+        self.gradient = {}
+
+        self.infeasible_side = self.response_settings["infeasible_side"].GetBool()
+        self.exponent = 2
+
+    @classmethod
+    def GetDefaultSettings(cls):
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:64]: What is this function used for?

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+            "model_import_settings" : {
+                "input_type"        : "use_input_model_part",
+                "input_filename"    : "UNKNOWN_NAME"
+            },
+            "infeasible_side"       : false
+        }""")
+        return this_defaults
+
+    def Initialize(self):
+        if self.model_part_needs_to_be_imported:
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]:
```suggestion
        if self.response_settings["model_import_settings"]["input_type"].GetString() == "mdpa":
```
This way you would not need the extra member variable `self.model_part_needs_to_be_imported`

**Review by philbucher — COMMENTED** [review]
do you need the `__init__.py`?
What is the problem ifyou don't use it?

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 26]
```diff
@@ -0,0 +1,157 @@
...
+    into a single response function.
+    The agglomeration happens by summing up the square of each nodal violation.
+    Nodes that are feasible do NOT contribute to the response value/gradient.
+    This is why a prediction of the violation using the gradients is not possible,
+    only correction of violations (e.g. from the last step) will happen.
+
+    Derived classes need to implement the calculation of the nodal violations
+
+    Important settings:
+    infeasible_side : boolean flag that indicates if the normal of bounding instance
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: what about naming this setting a bit more descriptive, like `infeasible_in_normal_direction`

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 27]
```diff
@@ -0,0 +1,157 @@
...
+
+    def GetValue(self):
+        return self.value
+
+    def GetShapeGradient(self):
+        if not self.gradient:
+            raise RuntimeError("Gradient was not calculated")
+        return self.gradient
+
+    def _CalculateProjectedDistances(self):
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: Also with those functions, I would remove the term "Projected" as it does not provide any useful additional info. How the distances are computed is an implementation detail.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py [hunk 28]
```diff
@@ -0,0 +1,157 @@
...
+        raise NotImplementedError("_CalculateProjectedDistances needs to be implemented by the derived class!")
+
+    def _CalculateValueForNode(self, signed_distance, direction):
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py]: What about the following names for this function and the one below:

`_CalculateNodalValue`
`_CalculateNodalGradient`

Your call.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py [hunk 29]
```diff
@@ -0,0 +1,63 @@
...
+#  KratosShapeOptimizationApplication
+#
+#  License:         BSD License
+#                   license: ShapeOptimizationApplication/license.txt
+#
+#  Main authors:    Geiser Armin, https://github.com/armingeiser
+#
+# ==============================================================================
+import KratosMultiphysics as KM
+
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py]: Remove empty line

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py [hunk 30]
```diff
@@ -0,0 +1,63 @@
...
+from .packaging_response_base import PackagingResponseBase
+from .. import custom_math as cm
+
+
+class PlanePackagingResponse(PackagingResponseBase):
+    """
+    A class for plane packaging response function. The plane is defined by an origin point and a normal vector.
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py]: Imho as a non-native speaker, the first sentence doesn't sound very English. 

```suggestion
    A class that defines the response function for plane-based packaging. The plane is defined by an origin point and a normal vector.
```
The change should be analogously done for the Mesh-based packaging.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py [hunk 31]
```diff
@@ -0,0 +1,63 @@
...
+#  Main authors:    Geiser Armin, https://github.com/armingeiser
+#
+# ==============================================================================
+import KratosMultiphysics as KM
+
+from .packaging_response_base import PackagingResponseBase
+from .. import custom_math as cm
+
+
+class PlanePackagingResponse(PackagingResponseBase):
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/plane_packaging_response.py]: See the comment above about the name of the response function.

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function_factory.py [hunk 32]
```diff
@@ -0,0 +1,23 @@
...
+def CreateResponseFunction(response_id, response_settings, model):
+    response_type = response_settings["response_type"].GetString()
+
+    if response_type == "plane_packaging":
+        return plane_packaging_response.PlanePackagingResponse(response_id, response_settings, model)
+    elif response_type == "mesh_packaging":
+        return mesh_packaging_response.MeshPackagingResponse(response_id, response_settings, model)
+    else:
+        raise NameError("The type of the following response function is not specified: "+ response_id +
+                        ".\nAvailable types are: 'plane_packaging'.")
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/response_functions/response_function_factory.py]:
```suggestion
                        ".\nAvailable types are: 'plane_packaging','mesh_packaging'.")
```
or

```suggestion
                        ".\nAvailable types are: 'plane_based_packaging','mesh_based_packaging'.")
```

File: applications/ShapeOptimizationApplication/tests/mesh_packaging_test/run_test.py [hunk 33]
```diff
@@ -0,0 +1,110 @@
...
+            gradient = {}
+            for node in current_design.Nodes:
+                if node.Id == node_id:
+                    gradient[node.Id] = [0.0, 0.0, 2*(node.Z-target)]
+                else:
+                    gradient[node.Id] = [0.0, 0.0, 0.0]
+
+            communicator.reportGradient("z_distance", gradient)
+
+
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/mesh_packaging_test/run_test.py]: Remove extra line.

File: applications/ShapeOptimizationApplication/tests/mesh_packaging_test/run_test.py [hunk 34]
```diff
@@ -0,0 +1,110 @@
...
+# Read parameters
+with open("optimization_parameters.json",'r') as parameter_file:
+    parameters = KM.Parameters(parameter_file.read())
+
+model = KM.Model()
+
+class CustomAnalyzer(AnalyzerBaseClass):
+    def AnalyzeDesignAndReportToCommunicator(self, current_design, optimization_iteration, communicator):
+
+        # Constraint 1
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/mesh_packaging_test/run_test.py]: Remove comment and move the code directly after the function declaration.

File: applications/ShapeOptimizationApplication/tests/plane_packaging_test/run_test.py [hunk 35]
```diff
@@ -0,0 +1,110 @@
...
+# Read parameters
+with open("optimization_parameters.json",'r') as parameter_file:
+    parameters = KM.Parameters(parameter_file.read())
+
+model = KM.Model()
+
+class CustomAnalyzer(AnalyzerBaseClass):
+    def AnalyzeDesignAndReportToCommunicator(self, current_design, optimization_iteration, communicator):
+
+        # Constraint 1
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/plane_packaging_test/run_test.py]: Remove comment and move the code directly after the function declaration.

File: applications/ShapeOptimizationApplication/tests/plane_packaging_test/run_test.py [hunk 36]
```diff
@@ -0,0 +1,110 @@
...
+        if communicator.isRequestingGradientOf("z_distance"):
+            gradient = {}
+            for node in current_design.Nodes:
+                if node.Id == node_id:
+                    gradient[node.Id] = [0.0, 0.0, 2*(node.Z-target)]
+                else:
+                    gradient[node.Id] = [0.0, 0.0, 0.0]
+
+            communicator.reportGradient("z_distance", gradient)
+
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/tests/plane_packaging_test/run_test.py:38]: remove extra line

**Review by dbaumgaertner — CHANGES_REQUESTED** [review]
Very very nice implementation! I also like having now again a generic ResponseFactory in the ShapeOptApp. It makes it easier to implement own ideas that are independent of the Kratos solvers, I guess. Thanks a lot, @armingeiser. 

Just some comparatively minor things. Mostly about naming.

**Comment by armingeiser** [comment]
`do you need the __init__.py?
What is the problem if you don't use it?`

@philbucher  The __init__.py file is needed for relative imports. With python 3.3+ it is not required anymore though
https://stackoverflow.com/questions/37139786/is-init-py-not-required-for-packages-in-python-3-3

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h (hunk 18 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]: done

File: applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h (hunk 19 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_utilities/geometry_utilities.h]: done

File: applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py (hunk 24 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/python_scripts/response_functions/packaging_response_base.py:64]: To get the default settings? it is a classmethod because they do not depend on the state of the object (self is not used)

**Comment by dbaumgaertner** [comment]
@armingeiser one more thing: In the following function, the last argument is not needed:

`def _CalculateValueForNode(self, signed_distance, direction):`

**Review by dbaumgaertner — APPROVED** [review]
Nice!

**Comment by dbaumgaertner** [comment]
@armingeiser Maybe just quickly check again the codacy comments. I think a few are meaningful here.
